### PR TITLE
Workflow Task Completion Metadata

### PR DESCRIPTION
**Pull request by Sushisource**
[Rendered](https://github.com/Sushisource/proposals/blob/wft-complete-metadata/api/wft-complete-metadata.md)

### CODE REVIEW

File: api/wft-complete-metadata.md
Comment (bergundy) [inline api/wft-complete-metadata.md:24]: I don't think we want to add this type of data.
We already know which activities and timers are unresolved from workflow history.
We don't know what the workflow code is effectively blocked on and we can't know that.

File: api/wft-complete-metadata.md
Comment (bergundy) [inline api/wft-complete-metadata.md:44]: I think this is a nice feature to have but I don't think this is the right mechanism for that since we need a tighter contract for the server to be able to interpret these config changes.

File: api/wft-complete-metadata.md
Comment (bergundy) [inline api/wft-complete-metadata.md:55]: Don't we have this information already?

File: api/wft-complete-metadata.md
Comment (bergundy) [inline api/wft-complete-metadata.md:83]: It really depends how this data is made visible in the UI.
I think struct is probably more than we'd need and is efficient enough.

File: api/wft-complete-metadata.md
Comment (bergundy) [inline api/wft-complete-metadata.md:109]: We're using this mechanism for versioning the SDK logic which effectively means it's for recreating workflow state.

File: api/wft-complete-metadata.md
Comment (bergundy) [inline api/wft-complete-metadata.md:114]: I just argued that above :)

I don't think the distinction matters.
To me it's more about leaving less traces when we make changes, especially if we charge users per marker generated in Cloud.

File: api/wft-complete-metadata.md
Comment (bergundy) [inline api/wft-complete-metadata.md:126]: This could imply slightly increased storage costs but I'd argue those are negligible.

File: api/wft-complete-metadata.md
Comment (Quinn-With-Two-Ns) [inline api/wft-complete-metadata.md:5]: For SDK internal versioning won't this feature effect replay/reconstitution of workfow state? Might be clearer to say user workfow state?

File: api/wft-complete-metadata.md
Comment (bergundy) [inline api/wft-complete-metadata.md:130]: I think we should definitely show this in the UI and make sure that the additional information is clearly visible.

File: api/wft-complete-metadata.md
Comment (cretz) [reply api/wft-complete-metadata.md:24]: Same. But I consider almost all of this as "could use metadata for one day" and not necessarily now.

File: api/wft-complete-metadata.md
Comment (cretz) [inline api/wft-complete-metadata.md:84]:
```suggestion
message RespondWorkflowTaskCompletedRequest {
  // Other fields

  map<string, temporal.api.common.v1.Payload> metadata = 11;
}

message WorkflowTaskCompletedEventAttributes {
  // Other fields

  map<string, temporal.api.common.v1.Payload> metadata = 6;
}
```

I think this has over-thought the issue. I think we can just use normal payloads and have reserved field names for ourselves. Our UI knows how to show user-structured payloads already and our codecs understand that payloads should be encrypted and users may want to encrypt this information.

If we wanted to keep the `StructuredMetadata` wrapper message, that's fine too. But I think we need to reuse our existing structured metadata mechanisms for structured metadata. We can discuss if certain keys should use the default converters (i.e. unencrypted).

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:24]: How would we not want to add this? This is incredibly useful for a user trying to understand what their workflow is currently waiting on.

Local activities are the prime example, right there, and they are *not* already in history. What rationale is there for not adding this? 

It's also not true that we don't know what code is blocked on. We know exactly what it's blocked on, because that's the definition of when you finish an activation / workflow task - when all routines are at an await point-type thing. They don't all necessarily correspond 1:1 to commands, but you know exactly what you're blocked on, by definition.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:83]: For things that you just want to show in the UI, sure. For notes the SDK wants to make to itself, a defined proto makes more sense.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:114]: Heh, I need to make more "Read the whole thing before commenting" notes at the top. 

You don't - I do. These sorts of rules don't make a difference in any specific situation much, but over the long run they make the difference between maintainable, sensible systems, and hodge-podge.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:5]: Yeah, which is what I get at later on at the bottom.

File: api/wft-complete-metadata.md
Comment (bergundy) [reply api/wft-complete-metadata.md:24]: What is this code blocked on?

```ts
export async function myWorkflow() {
  // run in the background and ignore if workflow is completed
  someActivity();
  await CancellationScope.current().cancelRequested;
}
```

You can tell by the stack trace (especially the structured one where users can see the actual code) but the SDK would struggle knowing which promises are "blocking" the workflow from making progress.

Local activities are definitely more interesting here since they don't show up in history until completed but even then I think we can infer from the fact that a WFT heartbeat is present that local activities are running and instruct the user to go to the stack trace query to get more information.

File: api/wft-complete-metadata.md
Comment (bergundy) [reply api/wft-complete-metadata.md:55]: All of the history events have timestamps, is that not enough? Do you mean something else here?

File: api/wft-complete-metadata.md
Comment (bergundy) [reply api/wft-complete-metadata.md:83]: Even then I'd want the information to be clearly displayed in the UI.

File: api/wft-complete-metadata.md
Comment (bergundy) [reply api/wft-complete-metadata.md:84]: I actually like the use of structured data here assuming that this data doesn't need to be encrypted.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:24]: It's blocked on waiting for the current cancellation scope to be requested. It's right there, where you're awaiting.

Admittedly that's harder to determine in TS than some other languages, but it's still 100% possible and useful. In Rust it's trivial. In Python it's probably pretty easy too.

Why would I want to _infer_ that we're probably waiting on some local activities when I could know exactly how many are currently running and their types? I don't really follow what the argument is for _not_ doing it? Besides that, we need this for cloud, which is one of the motivating reasons.

Comment (bergundy) [reply api/wft-complete-metadata.md:24]: My point is that we already know this from the stack trace, no need to add this data to histories.

Comment (Sushisource) [reply api/wft-complete-metadata.md:24]: Stack trace is a point-in-time view. This is a progressive view that works for every WFT, not just wherever you currently are.

A good point is that "currently waiting on" and "currently outstanding" are worth distinguishing. We need to know currently outstanding LAs for cloud, and perhaps things that are being waiting on is fine to leave for stack traces, but seems to me this would still be remarkably useful for debugging the progress of a workflow and it's not going to take up much space in history.

File: api/wft-complete-metadata.md
Comment (mmcshane) [reply api/wft-complete-metadata.md:84]: Having both the Any and the Struct maps isn't great - it's hard to predict or document where a given field will be as semantically they're the same. But we're really hamstrung by the protobuf type system here so the options aren't great. Do we need the two layers of lookup provided by a map of structs? Since a struct is itself a map it feels like we could get away with just a single Struct instance. What do you think about a convention where a `meta` field is always a Struct and if the Value of one of the Struct fields has to be a Payload in the future then so be it.

Comment (Sushisource) [reply api/wft-complete-metadata.md:84]: I'd be fine with just one `Struct` which can potentially have `Payload` (or `Any`) values in it.

Comment (cretz) [reply api/wft-complete-metadata.md:84]: I still think payload works very well for structured data and the UI knows how to deal with structured payload data and you get the bonus of encryptability and easy user value decoding. There are so many wins to using the same thing we use for memo, failure attributes, etc.

Comment (Sushisource) [reply api/wft-complete-metadata.md:84]: It works... OK. Something like `Struct` internally probably would've made more sense in them as our default for things we want renderable in the UI rather than serialized JSON which is a bit wasteful, but that's in the past so w/e. 

I take your point about just being able to re-use it.

Honestly, none of this is terribly consequential. They're all equivalent and it's just a matter of what's easier in what situation. I think we should just pick something and move on. I like `Struct` because it sets us up to easily and efficiently represent what I think will be the case with all or the vast majority of the data represented here, which is data that won't ever be passed through a custom codec. If we'd rather just keep using Payload b/c it exists, I get that too. @bergundy if you can tie break, I'd say let's do that and move forward (or maybe we can grab someone from server team)

Comment (bergundy) [reply api/wft-complete-metadata.md:84]: The main question is if we want users to be able to put their (potentially sensitive) data in these fields.
If that's not a requirement, we could go with Structs which are easier for us to work with and don't require extra serialization. If we're unsure, I'd go with @cretz's suggestion of using `Payload`s.

Comment (cretz) [reply api/wft-complete-metadata.md:84]: It's not just sensitive data. You're missing the other benefits I mentioned. What about the ability to easily decode to language-defined types in all languages? What about the consistency with all other metadata forms we use?

Comment (bergundy) [reply api/wft-complete-metadata.md:84]: I'm convinced.

Comment (Sushisource) [reply api/wft-complete-metadata.md:84]: OK, let's go with Payloads then.

I just want to say though that
> There just isn't a good reason for untyped structs

This isn't true - there is a pretty good reason, which is serializing and encoding JSON strings is wasteful by comparison, and they are easier to work with for clients who don't have access to all our payload machinery. Now, I'm fine with saying that's not worth re-using all the payloads stuff for this case - but that doesn't mean there's literally no merit to the idea.

Comment (cretz) [reply api/wft-complete-metadata.md:84]: > when we have a format that is perfect for this

Agree there is value in structs and other WKT for proto. For example Envoy uses proto to define their config formats and accepts these types for unknown user input instead of JSON.

IME though, working with these types in code makes you long for the typed protos or language structs you use otherwise. I've seen people just convert these things to JSON just so they can convert them into their language-typed forms anyways. But yes, they do have generic value.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:44]: I think that's probably true, but it's an option.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:55]: It could be anything arbitrarily more granular than that as well.

**Review by cretz — APPROVED** [review]

File: api/wft-complete-metadata.md
Comment (MichaelSnowden) [inline api/wft-complete-metadata.md:33]: Could you explain more about the changes mentioned here? When do we want the SDK to drive how the server writes history?

File: api/wft-complete-metadata.md
Comment (MichaelSnowden) [inline api/wft-complete-metadata.md:40]: I think we should remove this note since it's not really an acceptable option

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:33]: It won't affect how server writes history. It will affect how the SDK decides what to do, based on what version of an SDK a workflow was processed with, enabling or disabling certain features. SDKs may implement new features that affect how they produce commands, and on replay we need to do the same things that happened when the WF was first executed, hence a need to check our own version.

File: api/wft-complete-metadata.md
Comment (MichaelSnowden) [inline api/wft-complete-metadata.md:42]: I don't see the use case for making dynamic config adjustments in the WFT completion event. There are already existing methods to modify the dynamic config. Is the idea here that you can make it more atomic somehow? If so, that's something users may end up relying on. So we'd need to maintain that contract on the server-side, which seems error-prone to me and difficult to future-proof.

File: api/wft-complete-metadata.md
Comment (MichaelSnowden) [inline api/wft-complete-metadata.md:53]: In my experience, the span id is attached to the request metadata, not stored within the request data itself. What's the rationale behind moving this into the request data?

File: api/wft-complete-metadata.md
Comment (MichaelSnowden) [inline api/wft-complete-metadata.md:56]: This is already done via tracing, so it seems like a duplicate of the note above

File: api/wft-complete-metadata.md
Comment (MichaelSnowden) [inline api/wft-complete-metadata.md:57]: Spans already allow you to store arbitrary key/value pairs. Why can't we just use that?

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:42]: Maybe it wasn't clear, but these are all _possible_ use cases. The only one we know we need now is the SDK internal versioning ability, and probably the ability to track outstanding commands. Everything else is in "could be supported by this if we want" territory. Same applies for your subsequent comments, but I'll mention my thoughts on those too.

As mentioned here https://github.com/temporalio/proposals/pull/73#discussion_r1068742307 - yeah, maybe this one we'd avoid.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:53]: It's a simple matter of making it easily discoverable by a user looking at workflow history in the UI, who may very well not have initiated any of the requests associated with the workflow.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:56]: Fair enough.

File: api/wft-complete-metadata.md
Comment (MichaelSnowden) [inline api/wft-complete-metadata.md:69]: Adding arbitrarily-sized attributes into history exposes us to resource exhaustion. For example, we had a major outage caused by storing too many child workflow executions in the mutable state for executions.

File: api/wft-complete-metadata.md
Comment (MichaelSnowden) [inline api/wft-complete-metadata.md:92]: I think that simple arbitrary data associated with WFTs can be accomplished easily with tracing, and more advanced use cases should be left up to users. They can store their own data associated with a WFT id.

**Review by MichaelSnowden — COMMENTED** [review]
To continue with a review, I need to better understand the specific problem here. The solution here is very generic, but the only concrete use case I see is getting the version of the SDK, which produced a specific WFT, and I'm wondering why we even need that.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:69]: I resolved the other comments since they all largely resolve around the same concern w/ tracing.

Yep, bloat is called out in possible drawbacks. 

Not everyone uses tracing (in fact, I'd guess most users don't). This potentially gives them a way to write data into a place that we currently provide visibility most easily to them. Not everything makes sense to write as tracing data either.

Beyond that, this proposal isn't saying we'll definitely do that. It's something it could support if we ever needed to, now. No user data is going to be written here until if and when we decide we want to do that. If we did, almost certainly we'd put limits on that like we do w/ existing user-created fields.

File: api/wft-complete-metadata.md
Comment (Sushisource) [reply api/wft-complete-metadata.md:33]: See also https://github.com/temporalio/features/issues/200

File: api/wft-complete-metadata.md
Comment (yiminc) [inline api/wft-complete-metadata.md:28]: Will those payload be encoded by user's custom data converter? I would assume some of them would not if server needs to understand them, but others would because it could be sensitive info.

Comment (Sushisource) [reply api/wft-complete-metadata.md:28]: This specific example of outstanding commands should use the standard data converter. They may need to be understood for cloud billing purposes (the local activity stuff, depending on our choices there). They also need to be UI-renderable to be useful for debugging info.

But, regardless, again, that's not really required to move forward with this proposal. The specifics of any piece of metadata is up to if and when we add it. This proposal is just the mechanism of transport.

**Comment by Sushisource** [comment]
So, we talked more within the SDK team and we decided that the flexibility here comes at a bit too high of a cost in terms of size. We don't actually have any *concrete* use cases at the moment that require writing completely arbitrary data. Everything we want to write in here is well defined for at least one consumer (even if it's only the SDK consuming its own notes).

Hence, everything we know we want to add here is already a proto. The only question is how it will get encoded. (Hence the desire for `Any` in my original version). But, both `Any` and `Payload` are byte-wasteful.

The main motivation for adding this was speculation about needing to add generic stuff (but we don't know what that is) and decoupling SDK-only stuff from the long approval process / lifecycle in this repo. The first we can deal with if/when it comes up. The second we can deal with by simply introducing an `sdk` package in this repo with different codeowners. We'll do that.

**Pull request merged**
